[WIP] Updating Keepers Documentation (#478)
smartcontractkit/documentation: docs/chainlink-keepers/register-upkeep.md

title: 'Register Keeper Upkeep for a Contract'
whatsnext:
  {
    'FAQs': '/docs/chainlink-keepers/faqs/',
  }
---

# Overview <!-- omit in toc -->

> This guide will explain how to register a Keeper-compatible contract with the Chainlink Keeper Network. To find more information on deploying a Keeper-compatible contract, refer to [Making Compatible Contracts](../compatible-contracts). Registration of contracts can be completed using the Chainlink Keepers App:

Reviewer: ```suggestion
This guide explains how to register a Keeper-compatible contract with the Chainlink Keeper Network. To find more information about deploying a Keeper-compatible contract, see the [Making Compatible Contracts](../compatible-contracts) page. Register your contracts in the Chainlink Keepers App:
```

<div class="remix-callout">
    <a href="https://keepers.chain.link" class="cl-button--ghost solidity-tracked">Chainlink Keepers App</a>
</div>

After you register, you can interact directly with the [registry contract](https://etherscan.io/address/0x7b3EC232b08BD7b4b3305BE0C044D907B2DF960B#code) functions such as `cancelUpkeep` and `addFunds`.

**Table of Contents**
+ [Register Contract](#register-contract)
+ [Fund Upkeep](#fund-upkeep)
+ [How Funding Works](#how-funding-works)
+ [Maintain a Minimum Balance](#maintain-a-minimum-balance)
+ [Registration Completion](#registration-completion)
+ [Useful Patterns](#useful-patterns)
  + [Gas Limits](#gas-limits)
  + [Testing](#testing)

# Register Contract

> Registering an Upkeep with the Chainlink Keepers App notifies the Keeper Network about your contract, and allows you to fund it so your work is performed continuously. As part of the registration, we’re requesting some information that will help us to deliver the optimal experience for your use case as we continue to improve the product.

Reviewer: ```suggestion
Registering an Upkeep with the Chainlink Keepers App notifies the Keeper Network about your contract and allows you to fund it so your work is performed continuously. As part of the registration, we’re requesting some information that will help us to deliver the optimal experience for your use case as we continue to improve the product.
```

     Make sure you have LINK to fund your Upkeep with. Learn how to [acquire testnet LINK](/docs/acquire-link/).

    > ❗️ Funding Upkeep
    > You should fund your contract with more LINK that you anticipate you will need. The network will not check or perform your Upkeep if your balance could be too low based on current exchange rates.
    >
    > Your balance will be charged LINK based on a 20% premium over the gas cost to `performUpkeep`. Gas costs include the gas required for your Keeper-compatible contract to complete execution and an 80k overhead from the `KeeperRegistry` itself. The premium and overhead are not fixed and will change over time.

    The gas limit of the example counter contract should be set to 200,000.

> 2. **Press `Register upkeep`** and confirm the transaction in MetaMask

Reviewer: ```suggestion
1. **Click `Register upkeep`** and confirm the transaction in MetaMask
```

If this isn't auto-ordering, you might need to adjust the tabbing.

  * Receive a success message and verify that the funds were added to the Upkeep
    ![Funds Added Successful Message](/images/contract-devs/keeper/keeper-add-funds.png)

# How Funding Works

* Your balance is reduced each time a Keeper executes your `performUpkeep` method.
* There is no cost for `checkUpkeep` calls.
* If you want to automate adding funds, you can directly call the `addFunds()` function on the `KeeperRegistry` contract.
* Anyone can call the `addFunds()` function, not just the Upkeep owner.
* To withdraw funds, cancel the Upkeep.

# Maintain a Minimum Balance
To ensure that the Chainlink Keepers are compensated for performance, there is an expected minimum balance on each Upkeep. If your funds drop below this amount, the Upkeep will not be performed.

The minimum balance is calculated using the current fast gas price, the Gas Limit you entered for your Upkeep, and the max gas multiplier (see `gasCeilingMultiplier` in [configuration of the registry](../overview/#configuration)).

It is recommended that you maintain a balance that is a multiple (3-5x) of the minimum balance to account for gas price fluctuations.

# Registration Completion

> After you register your Upkeep and it has been approved with sufficient funds, the Chainlink Keeper Network will begin to simulate `checkUpkeep` calls and execute your contract's `performUpkeep` function as needed.

Reviewer: ```suggestion
After you register your Upkeep and it is approved with sufficient funds, the Chainlink Keeper Network starts to simulate `checkUpkeep` calls and execute your contract's `performUpkeep` function as needed.
```

You have now successfully built and registered a Keeper Compatible contract with the Chainlink Keeper Network.

# Useful Patterns

These patterns are not mutually exclusive. Review and make use of the patterns that make sense for your use case.

## Gas Limits

> ❗️ Gas Limits
>
> The `KeeperRegistry` enforces a cap for gas used both on-chain and off-chain. See the [Keepers Network Overview](../overview/) for details. The caps are configurable and might change based on user feedback. Be sure that you understand these limits if your use case requires a large amount of gas.

> When developing your keeper-compatible smart contracts, it is critical to understand the gas limits you are working with on the KeeperRegistry. There is a `check` gas limit and a `call` gas limit that your contract must adhere to in order to operate successfully. See the [Keepers Network Overview](../overview/) to learn the current configuration.

Reviewer: ```suggestion
When developing your keeper-compatible smart contracts, you must understand the gas limits that you are working with on the KeeperRegistry. There is a `check` gas limit and a `call` gas limit that your contract must adhere to in order to operate successfully. See the [Keepers Network Overview](../overview/) to learn the current configuration.
```
